toolchain/check/function.cpp

```diff
@@ -0,0 +1,216 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/function.h"
+
+namespace Carbon::Check {
+
+// Returns true if the function signature has an error, which will have
+// previously been diagnosed.
+static auto FunctionSignatureHasError(Context& context,
+                                      const SemIR::Function& fn) -> bool {
+  if (fn.return_type_id == SemIR::TypeId::Error) {
+    return true;
+  }
+  for (auto param_refs_id : {fn.implicit_param_refs_id, fn.param_refs_id}) {
+    if (param_refs_id != SemIR::InstBlockId::Empty) {
+      for (auto param_id : context.inst_blocks().Get(param_refs_id)) {
+        if (context.insts().Get(param_id).type_id() == SemIR::TypeId::Error) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+// Returns true if a param agrees. The caller is expected to provide a
+// diagnostic.
+// TODO: Consider moving diagnostics here, particularly to differentiate
+// between type and name mistakes. For now, taking the simpler approach because
+// I also think we may want to refactor params.
+static auto DoesParamAgree(Context& context, SemIR::InstId new_param_ref_id,
+                           SemIR::InstId prev_param_ref_id) -> bool {
+  auto new_param_ref = context.insts().Get(new_param_ref_id);
+  auto prev_param_ref = context.insts().Get(prev_param_ref_id);
+  if (new_param_ref.kind() != prev_param_ref.kind() ||
+      new_param_ref.type_id() != prev_param_ref.type_id()) {
+    return false;
+  }
+
+  if (new_param_ref.Is<SemIR::AddrPattern>()) {
+    new_param_ref =
+        context.insts().Get(new_param_ref.As<SemIR::AddrPattern>().inner_id);
+    prev_param_ref =
+        context.insts().Get(prev_param_ref.As<SemIR::AddrPattern>().inner_id);
+  }
+
+  if (new_param_ref.kind() != prev_param_ref.kind()) {
+    return false;
+  }
+
+  if (new_param_ref.Is<SemIR::AnyBindName>()) {
+    new_param_ref =
+        context.insts().Get(new_param_ref.As<SemIR::AnyBindName>().value_id);
+    prev_param_ref =
+        context.insts().Get(prev_param_ref.As<SemIR::AnyBindName>().value_id);
+  }
+
+  CARBON_CHECK(new_param_ref.Is<SemIR::Param>()) << new_param_ref;
+  CARBON_CHECK(prev_param_ref.Is<SemIR::Param>()) << prev_param_ref;
+
+  auto new_param = new_param_ref.As<SemIR::Param>();
+  auto prev_param = prev_param_ref.As<SemIR::Param>();
+  return new_param.name_id == prev_param.name_id;
+}
+
+// Returns true if two param refs agree.
+static auto DoParamAgrees(Context& context, SemIR::InstId new_decl_id,
```

Comment:
```suggestion
static auto DoParamsAgree(Context& context, SemIR::InstId new_decl_id,
```

I think? (I'd read "DoParamAgrees" as "do the 'param-agrees' process", which might be accurate but seems a bit awkward.)

But given that this is producing diagnostics rather than just computing a `bool`, I think naming this so it doesn't sound like a predicate might be better. Maybe:

```suggestion
static auto CheckParamsAgree(Context& context, SemIR::InstId new_decl_id,
```

Reply:
This was some sloppy naming, but how about just `CheckRedecl` + `Param` + `s` as naming. :)

Reply:
That name works for me, but I think it'd be better to have more difference between the names of `CheckRedeclParams` and `CheckRedeclParam` to reflect that one of them is producing diagnostics and the other is just performing a comparison. (Or maybe make them both produce diagnostics per the FIXME.)

Reply:
Moved (not really addressing the FIXME, I don't think the structure is worth addressing)